crates/iceberg-ext/src/spec/spec_binder.rs

```diff
@@ -0,0 +1,476 @@
+use crate::catalog::rest::ErrorModel;
+use http::StatusCode;
+use iceberg::spec::{
+    NestedFieldRef, PartitionField, PartitionSpec, SchemaRef, Transform, Type,
+    UnboundPartitionField, UnboundPartitionSpec,
+};
+use std::cmp::max;
+use std::collections::HashSet;
+
+type Result<T> = std::result::Result<T, ErrorModel>;
+
+#[derive(Debug)]
+pub struct PartitionSpecBinder {
+    spec_id: i32,
+    schema: SchemaRef,
+    partition_names: HashSet<String>,
+    last_assigned_field_id: i32,
+}
+
+impl PartitionSpecBinder {
+    const PARTITION_DATA_ID_START: i32 = 1000;
+
+    pub fn new(schema: SchemaRef, spec_id: i32) -> Self {
+        Self {
+            spec_id,
+            schema,
+            last_assigned_field_id: Self::PARTITION_DATA_ID_START - 1,
+            partition_names: HashSet::new(),
+        }
+    }
+
+    pub fn bind(mut self, spec: &UnboundPartitionSpec) -> Result<PartitionSpec> {
+        Ok(PartitionSpec {
+            spec_id: self.spec_id,
+            fields: spec
+                .fields
+                .iter()
+                .map(|unbounded_field| self.bind_field(unbounded_field))
+                .collect::<Result<Vec<PartitionField>>>()?,
```

Comment:
There is a check missing here I believe: We need to make sure, that each PartitionField only occurs once in a Partition Spec. Java has the `CheckForRedundantPartitions` function (https://github.com/apache/iceberg/blob/2e291c2b67a643ffe93e139483df55f3639cc39d/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L393) which is then called by each of the transforms:
https://github.com/apache/iceberg/blob/2e291c2b67a643ffe93e139483df55f3639cc39d/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L423

We should also add a test for this then, to make sure that we don't have the same column twice in a single partition spec.